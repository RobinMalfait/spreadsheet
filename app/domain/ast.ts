import type { Location } from '~/domain/expression'

export enum AstKind {
  CELL = 'CELL',
  RANGE = 'RANGE',
  FUNCTION = 'FUNCTION',
  NUMBER_LITERAL = 'NUMBER_LITERAL',
  STRING_LITERAL = 'STRING_LITERAL',
  BINARY_EXPRESSION = 'BINARY_EXPRESSION',
}

interface Span {
  span: {
    start: number
    end: number
  }
}

export interface AstCell extends Span {
  kind: AstKind.CELL
  name: string
  raw: string
  loc: Location
}

export interface AstCellRange extends Span {
  kind: AstKind.RANGE
  raw: string
  start: AstCell
  end: AstCell
}

export interface AstFunction extends Span {
  kind: AstKind.FUNCTION
  name: string
  args: AST[]
}

interface AstNumberLiteral extends Span {
  kind: AstKind.NUMBER_LITERAL
  value: number
}

interface AstStringLiteral extends Span {
  kind: AstKind.STRING_LITERAL
  value: string
}

interface AstBinaryExpression extends Span {
  kind: AstKind.BINARY_EXPRESSION
  operator: BinaryExpressionOperator
  lhs: AST
  rhs: AST
}

export type AST =
  | AstCell
  | AstCellRange
  | AstFunction
  | AstNumberLiteral
  | AstStringLiteral
  | AstBinaryExpression

export enum BinaryExpressionOperator {
  // Math operators
  EXPONENT = 'EXPONENT',
  MULTIPLY = 'MULTIPLY',
  ADD = 'ADD',
  SUBTRACT = 'SUBTRACT',
  DIVIDE = 'DIVIDE',

  // Comparison operators
  EQUALS = 'EQUALS',
  NOT_EQUALS = 'NOT_EQUALS',
  LESS_THAN = 'LESS_THAN',
  LESS_THAN_EQUALS = 'LESS_THAN_EQUALS',
  GREATER_THAN = 'GREATER_THAN',
  GREATER_THAN_EQUALS = 'GREATER_THAN_EQUALS',
}
