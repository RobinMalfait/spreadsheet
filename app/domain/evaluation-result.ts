export enum EvaluationResultKind {
  ERROR = 'ERROR',
  EMPTY = 'EMPTY',
  NUMBER = 'NUMBER',
  STRING = 'STRING',
  BOOLEAN = 'BOOLEAN',
  DATETIME = 'DATETIME',
}

export type EvaluationResult =
  | { kind: EvaluationResultKind.ERROR; value: string }
  | { kind: EvaluationResultKind.EMPTY; value: string }
  | { kind: EvaluationResultKind.NUMBER; value: number }
  | { kind: EvaluationResultKind.STRING; value: string }
  | { kind: EvaluationResultKind.BOOLEAN; value: boolean }
  | { kind: EvaluationResultKind.DATETIME; value: Date; date: boolean; time: boolean }
