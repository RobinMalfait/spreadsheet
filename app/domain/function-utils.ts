import type { EvaluationResult } from '~/domain/evaluation-result'
import { parse } from '~/domain/signature/parser'
import { tokenize } from '~/domain/signature/tokenizer'
import { validate } from '~/domain/signature/validate'

export function expose<
  T extends EvaluationResult,
  R extends EvaluationResult | EvaluationResult[],
>(signature: string, handle: (...args: T[]) => R): (...args: T[]) => R {
  let sig = parse(tokenize(signature))

  return Object.assign(
    (...args: T[]) => {
      let errors = validate(sig, args)
      if (errors) return errors as R

      // Handle
      return handle(...args) as R
    },
    { signature: sig },
  )
}
