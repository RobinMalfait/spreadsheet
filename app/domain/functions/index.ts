export * from './binary'
export * from './date'
export * from './logic'
export * from './lookup'
export * from './math'
export * from './privileged'
export * from './sequence'
export * from './statistics'
export * from './text'
export * from './types'
