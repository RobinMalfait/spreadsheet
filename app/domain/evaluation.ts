import { format } from 'date-fns'
import { type AST, AstKind, BinaryExpressionOperator } from '~/domain/ast'
import * as functions from '~/domain/functions'
import type { Spreadsheet } from '~/domain/spreadsheet'
import { expandRange } from '~/domain/walk-ast'

export enum EvaluationResultKind {
  NUMBER = 'NUMBER',
  STRING = 'STRING',
  BOOLEAN = 'BOOLEAN',
  DATETIME = 'DATETIME',
}

export type EvaluationResult =
  | { kind: EvaluationResultKind.NUMBER; value: number }
  | { kind: EvaluationResultKind.STRING; value: string }
  | { kind: EvaluationResultKind.BOOLEAN; value: boolean }
  | { kind: EvaluationResultKind.DATETIME; value: Date; date: boolean; time: boolean }

export function evaluateExpression(
  ast: AST,
  spreadsheet: Spreadsheet,
): EvaluationResult[] {
  switch (ast.kind) {
    case AstKind.NUMBER_LITERAL:
      return [{ kind: EvaluationResultKind.NUMBER, value: ast.value }]

    case AstKind.STRING_LITERAL: {
      if (ast.value.trim() === '') {
        return [{ kind: EvaluationResultKind.STRING, value: ast.value }]
      }

      // Try to coerce the string to a number, and if it works, return a number
      // instead of a string
      let asNumber = Number(ast.value)
      if (!Number.isNaN(asNumber)) {
        return [{ kind: EvaluationResultKind.NUMBER, value: asNumber }]
      }

      return [{ kind: EvaluationResultKind.STRING, value: ast.value }]
    }

    case AstKind.CELL:
      return spreadsheet.evaluate(ast.name)

    case AstKind.RANGE: {
      let out = []
      for (let cell of expandRange(ast)) {
        for (let child of spreadsheet.evaluate(cell)) {
          out.push(child)
        }
      }
      return out
    }

    case AstKind.BINARY_EXPRESSION: {
      let lhs = evaluateExpression(ast.lhs, spreadsheet)
      let rhs = evaluateExpression(ast.rhs, spreadsheet)

      if (lhs.length !== 1 || rhs.length !== 1) {
        throw new Error('Expected a single result from each side')
      }

      let left = lhs[0]
      let right = rhs[0]

      if (
        left?.kind === EvaluationResultKind.NUMBER &&
        right?.kind === EvaluationResultKind.NUMBER
      ) {
        switch (ast.operator) {
          // Math operators
          case BinaryExpressionOperator.EXPONENT:
            return [functions.POWER(left, right)]
          case BinaryExpressionOperator.MULTIPLY:
            return [functions.MULTIPLY(left, right)]
          case BinaryExpressionOperator.DIVIDE:
            return [functions.DIVIDE(left, right)]
          case BinaryExpressionOperator.ADD:
            return [functions.SUM(left, right)]
          case BinaryExpressionOperator.SUBTRACT:
            return [functions.SUBTRACT(left, right)]

          // Comparison operators
          case BinaryExpressionOperator.EQUALS:
            return left.value === right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.NOT_EQUALS:
            return left.value !== right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.LESS_THAN:
            return left.value < right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.LESS_THAN_EQUALS:
            return left.value <= right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.GREATER_THAN:
            return left.value > right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.GREATER_THAN_EQUALS:
            return left.value >= right.value ? [functions.TRUE()] : [functions.FALSE()]
          default:
            ast.operator satisfies never
        }
      }

      throw new Error(`Invalid operation \`${ast.kind}\``)
    }

    case AstKind.FUNCTION: {
      if (!Object.hasOwn(functions, ast.name)) {
        throw new Error(`Unknown function \`${ast.name}\``)
      }

      let fn = functions[ast.name as keyof typeof functions]
      let args = ast.args.flatMap((arg) => evaluateExpression(arg, spreadsheet))
      // @ts-expect-error Some functions have a different arity, but we're not
      // checking that here.
      let result = fn(...args)
      if (Array.isArray(result)) {
        return result
      }
      return [result]
    }
  }
}

export function printEvaluationResult(result: EvaluationResult): string {
  switch (result.kind) {
    case EvaluationResultKind.NUMBER:
      return result.value.toString()
    case EvaluationResultKind.STRING:
      return result.value
    case EvaluationResultKind.BOOLEAN:
      return result.value ? 'TRUE' : 'FALSE'
    case EvaluationResultKind.DATETIME:
      if (result.date && result.time) return format(result.value, 'yyyy-MM-dd HH:mm:ss')
      if (result.date) return format(result.value, 'yyyy-MM-dd')
      return format(result.value, 'HH:mm:ss')
    default:
      result satisfies never
      return ''
  }
}
