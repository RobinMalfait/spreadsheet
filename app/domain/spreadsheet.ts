import { type AST, AstKind } from '~/domain/ast'
import { type EvaluationResult, evaluateExpression } from '~/domain/evaluation'
import { parse } from '~/domain/expression'
import * as functions from '~/domain/functions'
import { tokenize } from '~/domain/tokenizer'
import { WalkAction, walk } from '~/domain/walk-ast'
import { DefaultMap } from '~/utils/default-map'

export enum ComputationResultKind {
  VALUE = 'VALUE',
  ERROR = 'ERROR',
}

type ComputationValue = {
  kind: ComputationResultKind.VALUE
  value: EvaluationResult
}

type ComputationError = {
  kind: ComputationResultKind.ERROR
  short: string
  message: string
}

export class Spreadsheet {
  // Track each individual cell and it's contents. AST is pre-parsed.
  private cells = new Map<string, [raw: string, ast: AST]>()

  // Track all dependencies for each cell.
  private _dependencies = new DefaultMap<string, Set<string>>(() => new Set<string>())

  private evaluationCache = new Map<string, EvaluationResult[]>()

  get cellNames() {
    return Array.from(this.cells.keys())
  }

  has(cell: string): boolean {
    return this.cells.has(cell)
  }

  get(cell: string): string {
    let result = this.cells.get(cell)
    if (result) {
      return result[0]
    }
    return ''
  }

  set(cell: string, value: string) {
    // Reset state
    this.cells.delete(cell)

    // Clear existing dependencies for this cell
    let dependencies = this._dependencies.get(cell)
    dependencies.clear()

    // Clear the full evaluation cache. Very naive, but let's re-compute
    // everything the moment _something_ changes.
    this.evaluationCache.clear()

    // Value must be filled in
    if (value.trim() === '') return

    // Expression should exist
    let expression = value[0] === '=' ? value.slice(1) : `"${value}"`
    if (expression.trim() === '') return

    let tokens = tokenize(expression)
    let ast = parse(tokens)

    // Track all references in the AST
    walk([ast], (node) => {
      if (node.kind === AstKind.CELL) {
        dependencies.add(node.name)
      }

      return WalkAction.Continue
    })

    this.cells.set(cell, [value, ast])
  }

  functions() {
    let collator = new Intl.Collator('en', { sensitivity: 'base' })
    return Object.keys(functions).sort(collator.compare)
  }

  dependencies(cell: string): Set<string> {
    return new Set(this._dependencies.get(cell))
  }

  compute(cell: string): ComputationValue | ComputationError | null {
    try {
      let evaluationResult = this.evaluate(cell)
      if (evaluationResult.length > 1) {
        throw Object.assign(
          new Error(`Expected a single result, got ${evaluationResult.length}`),
          { short: '#VALUE!' },
        )
      }

      if (evaluationResult.length === 1) {
        return {
          kind: ComputationResultKind.VALUE,
          // biome-ignore lint/style/noNonNullAssertion: we already verified that the evaluationResult has at least one element.
          value: evaluationResult[0]!,
        }
      }

      return null
    } catch (err: unknown) {
      return {
        kind: ComputationResultKind.ERROR,
        // @ts-expect-error This is fineâ€¦
        short: err?.short ?? 'Error',
        message: (err as Error).message,
      }
    }
  }

  evaluate(cell: string): EvaluationResult[] {
    let result = this.cells.get(cell)
    if (!result) return []

    let cached = this.evaluationCache.get(cell)
    if (cached) return cached

    // TODO: Should this be moved to the `set` method?
    if (result[1].kind === AstKind.RANGE) {
      throw Object.assign(new Error('Cannot reference a range to a cell'), {
        short: '#VALUE!',
      })
    }

    // Verify references
    let dependencies = this._dependencies.get(cell)
    if (dependencies.size > 0) {
      let handled = new Set<string>()
      let todo = Array.from(dependencies)

      // Track where we came from
      let previous = cell

      while (todo.length > 0) {
        // biome-ignore lint/style/noNonNullAssertion: We verified that there is at least a single element in the while condition above.
        let next = todo.shift()!

        // No need to re-verify cells we've already handled
        if (handled.has(next)) continue
        handled.add(next)

        if (next === cell) {
          throw Object.assign(
            new Error(`Circular reference detected in cell ${previous}`),
            { short: '#REF!' },
          )
        }

        if (this.cells.has(next)) {
          // Track where we came from
          previous = next

          // Check transitive dependencies
          for (let other of this._dependencies.get(next)) {
            if (!handled.has(other)) {
              todo.push(other)
            }
          }
        }
      }
    }

    let out = evaluateExpression(result[1], this)
    this.evaluationCache.set(cell, out)
    return out
  }
}
