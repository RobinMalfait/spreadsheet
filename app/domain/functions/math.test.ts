import { describe, expect, it } from 'vitest'
import { Spreadsheet } from '~/domain/spreadsheet'
import { visualizeSpreadsheet } from '~/test/utils'

describe('PI()', () => {
  it('should result in the value of PI', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=PI()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────────────────┐
      │   │ A                 │
      ├───┼───────────────────┤
      │ 1 │ 3.141592653589793 │
      └───┴───────────────────┘
      "
    `)
  })
})

describe('TAU()', () => {
  it('should result in the value of TAU', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=TAU()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────────────────┐
      │   │ A                 │
      ├───┼───────────────────┤
      │ 1 │ 6.283185307179586 │
      └───┴───────────────────┘
      "
    `)
  })
})

describe('SUM()', () => {
  it('should sum up all the numbers', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=2')
    spreadsheet.set('B1', '=4')
    spreadsheet.set('C1', '=6')
    spreadsheet.set('A2', '=SUM(A1:C1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬────┬───┬───┐
      │   │ A  │ B │ C │
      ├───┼────┼───┼───┤
      │ 1 │ 2  │ 4 │ 6 │
      ├───┼────┼───┼───┤
      │ 2 │ 12 │   │   │
      └───┴────┴───┴───┘
      "
    `)
  })
})

describe('SUBTRACT()', () => {
  it('should subtract all the numbers', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=12')
    spreadsheet.set('B1', '=6')
    spreadsheet.set('C1', '=3')
    spreadsheet.set('A2', '=SUBTRACT(A1:C1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬────┬───┬───┐
      │   │ A  │ B │ C │
      ├───┼────┼───┼───┤
      │ 1 │ 12 │ 6 │ 3 │
      ├───┼────┼───┼───┤
      │ 2 │ 3  │   │   │
      └───┴────┴───┴───┘
      "
    `)
  })
})

describe('MULTIPLY()', () => {
  it('should multiply two numbers', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=2')
    spreadsheet.set('B1', '=3')
    spreadsheet.set('A2', '=PRODUCT(A1, B1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───┬───┐
      │   │ A │ B │
      ├───┼───┼───┤
      │ 1 │ 2 │ 3 │
      ├───┼───┼───┤
      │ 2 │ 6 │   │
      └───┴───┴───┘
      "
    `)
  })
})

describe('PRODUCT()', () => {
  it('should multiply all the numbers', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=2')
    spreadsheet.set('B1', '=3')
    spreadsheet.set('C1', '=4')
    spreadsheet.set('A2', '=PRODUCT(A1:C1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬────┬───┬───┐
      │   │ A  │ B │ C │
      ├───┼────┼───┼───┤
      │ 1 │ 2  │ 3 │ 4 │
      ├───┼────┼───┼───┤
      │ 2 │ 24 │   │   │
      └───┴────┴───┴───┘
      "
    `)
  })
})

describe('DIVIDE()', () => {
  it('should divide two numbers', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=12')
    spreadsheet.set('B1', '=2')
    spreadsheet.set('A2', '=DIVIDE(A1:B1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬────┬───┐
      │   │ A  │ B │
      ├───┼────┼───┤
      │ 1 │ 12 │ 2 │
      ├───┼────┼───┤
      │ 2 │ 6  │   │
      └───┴────┴───┘
      "
    `)
  })
})

describe('POWER()', () => {
  it('should calculate the power of a number', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=12')
    spreadsheet.set('B1', '=2')
    spreadsheet.set('A2', '=POWER(A1, B1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────┬───┐
      │   │ A   │ B │
      ├───┼─────┼───┤
      │ 1 │ 12  │ 2 │
      ├───┼─────┼───┤
      │ 2 │ 144 │   │
      └───┴─────┴───┘
      "
    `)
  })
})

describe('AVERAGE()', () => {
  it('should calculate the average of a set of numbers', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=2')
    spreadsheet.set('B1', '=3')
    spreadsheet.set('C1', '=4')
    spreadsheet.set('A2', '=AVERAGE(A1:C1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───┬───┬───┐
      │   │ A │ B │ C │
      ├───┼───┼───┼───┤
      │ 1 │ 2 │ 3 │ 4 │
      ├───┼───┼───┼───┤
      │ 2 │ 3 │   │   │
      └───┴───┴───┴───┘
      "
    `)
  })
})

describe('MOD()', () => {
  it('should compute the modulo', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=13')
    spreadsheet.set('B1', '=10')
    spreadsheet.set('A2', '=MOD(A1, B1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬────┬────┐
      │   │ A  │ B  │
      ├───┼────┼────┤
      │ 1 │ 13 │ 10 │
      ├───┼────┼────┤
      │ 2 │ 3  │    │
      └───┴────┴────┘
      "
    `)
  })
})

describe('FLOOR()', () => {
  it('should floor a number', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=PI()')
    spreadsheet.set('B1', '=FLOOR(A1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────────────────┬───┐
      │   │ A                 │ B │
      ├───┼───────────────────┼───┤
      │ 1 │ 3.141592653589793 │ 3 │
      └───┴───────────────────┴───┘
      "
    `)
  })
})

describe('CEIL()', () => {
  it('should ceil a number', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=PI()')
    spreadsheet.set('B1', '=CEIL(A1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────────────────┬───┐
      │   │ A                 │ B │
      ├───┼───────────────────┼───┤
      │ 1 │ 3.141592653589793 │ 4 │
      └───┴───────────────────┴───┘
      "
    `)
  })
})

describe('ROUND()', () => {
  it('should round a number', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=PI()')
    spreadsheet.set('B1', '=ROUND(A1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────────────────┬───┐
      │   │ A                 │ B │
      ├───┼───────────────────┼───┤
      │ 1 │ 3.141592653589793 │ 3 │
      └───┴───────────────────┴───┘
      "
    `)
  })

  it('should round a number with 1 decimal of precision', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=PI()')
    spreadsheet.set('B1', '=ROUND(A1, 1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────────────────┬─────┐
      │   │ A                 │ B   │
      ├───┼───────────────────┼─────┤
      │ 1 │ 3.141592653589793 │ 3.1 │
      └───┴───────────────────┴─────┘
      "
    `)
  })

  it('should round a number with 2 decimals of precision', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=PI()')
    spreadsheet.set('B1', '=ROUND(A1, 2)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────────────────┬──────┐
      │   │ A                 │ B    │
      ├───┼───────────────────┼──────┤
      │ 1 │ 3.141592653589793 │ 3.14 │
      └───┴───────────────────┴──────┘
      "
    `)
  })
})
