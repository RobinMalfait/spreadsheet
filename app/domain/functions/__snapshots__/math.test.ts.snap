// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Example tests > ABS(1) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 1 │
└───┴───┘
"
`;

exports[`Example tests > ACOS(1) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 0 │
└───┴───┘
"
`;

exports[`Example tests > ACOSH(1) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 0 │
└───┴───┘
"
`;

exports[`Example tests > ADD(1, 2) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 3 │
└───┴───┘
"
`;

exports[`Example tests > ASIN(1) 1`] = `
"
┌───┬────────────────────┐
│   │ A                  │
├───┼────────────────────┤
│ 1 │ 1.5707963267948966 │
└───┴────────────────────┘
"
`;

exports[`Example tests > ASINH(1) 1`] = `
"
┌───┬───────────────────┐
│   │ A                 │
├───┼───────────────────┤
│ 1 │ 0.881373587019543 │
└───┴───────────────────┘
"
`;

exports[`Example tests > ATAN(1) 1`] = `
"
┌───┬────────────────────┐
│   │ A                  │
├───┼────────────────────┤
│ 1 │ 0.7853981633974483 │
└───┴────────────────────┘
"
`;

exports[`Example tests > ATAN2(1, 1) 1`] = `
"
┌───┬────────────────────┐
│   │ A                  │
├───┼────────────────────┤
│ 1 │ 0.7853981633974483 │
└───┴────────────────────┘
"
`;

exports[`Example tests > ATANH(1) 1`] = `
"
┌───┬──────────┐
│   │ A        │
├───┼──────────┤
│ 1 │ Infinity │
└───┴──────────┘
"
`;

exports[`Example tests > CBRT(1) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 1 │
└───┴───┘
"
`;

exports[`Example tests > CEIL(1.5) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 2 │
└───┴───┘
"
`;

exports[`Example tests > CLZ32(1) 1`] = `
"
┌───┬────┐
│   │ A  │
├───┼────┤
│ 1 │ 31 │
└───┴────┘
"
`;

exports[`Example tests > COS(1) 1`] = `
"
┌───┬────────────────────┐
│   │ A                  │
├───┼────────────────────┤
│ 1 │ 0.5403023058681398 │
└───┴────────────────────┘
"
`;

exports[`Example tests > COSH(1) 1`] = `
"
┌───┬────────────────────┐
│   │ A                  │
├───┼────────────────────┤
│ 1 │ 1.5430806348152437 │
└───┴────────────────────┘
"
`;

exports[`Example tests > DIVIDE(6, 3) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 2 │
└───┴───┘
"
`;

exports[`Example tests > EXP(1) 1`] = `
"
┌───┬───────────────────┐
│   │ A                 │
├───┼───────────────────┤
│ 1 │ 2.718281828459045 │
└───┴───────────────────┘
"
`;

exports[`Example tests > FLOOR(1.5) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 1 │
└───┴───┘
"
`;

exports[`Example tests > IMUL(1, 2) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 2 │
└───┴───┘
"
`;

exports[`Example tests > LOG(1) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 0 │
└───┴───┘
"
`;

exports[`Example tests > LOG10(1) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 0 │
└───┴───┘
"
`;

exports[`Example tests > MOD(5, 2) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 1 │
└───┴───┘
"
`;

exports[`Example tests > MULTIPLY(2, 3) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 6 │
└───┴───┘
"
`;

exports[`Example tests > PI() 1`] = `
"
┌───┬───────────────────┐
│   │ A                 │
├───┼───────────────────┤
│ 1 │ 3.141592653589793 │
└───┴───────────────────┘
"
`;

exports[`Example tests > POWER(2, 3) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 8 │
└───┴───┘
"
`;

exports[`Example tests > PRODUCT(2, 3, 4) 1`] = `
"
┌───┬────┐
│   │ A  │
├───┼────┤
│ 1 │ 24 │
└───┴────┘
"
`;

exports[`Example tests > ROUND(1.5) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 2 │
└───┴───┘
"
`;

exports[`Example tests > ROUND(PI(), 2) 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ 3.14 │
└───┴──────┘
"
`;

exports[`Example tests > SIN(1) 1`] = `
"
┌───┬────────────────────┐
│   │ A                  │
├───┼────────────────────┤
│ 1 │ 0.8414709848078965 │
└───┴────────────────────┘
"
`;

exports[`Example tests > SINH(1) 1`] = `
"
┌───┬────────────────────┐
│   │ A                  │
├───┼────────────────────┤
│ 1 │ 1.1752011936438014 │
└───┴────────────────────┘
"
`;

exports[`Example tests > SQRT(1) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 1 │
└───┴───┘
"
`;

exports[`Example tests > SUBTRACT(2, 1) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 1 │
└───┴───┘
"
`;

exports[`Example tests > SUM(1, 2, 3) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 6 │
└───┴───┘
"
`;

exports[`Example tests > TAN(1) 1`] = `
"
┌───┬────────────────────┐
│   │ A                  │
├───┼────────────────────┤
│ 1 │ 1.5574077246549023 │
└───┴────────────────────┘
"
`;

exports[`Example tests > TANH(1) 1`] = `
"
┌───┬────────────────────┐
│   │ A                  │
├───┼────────────────────┤
│ 1 │ 0.7615941559557649 │
└───┴────────────────────┘
"
`;

exports[`Example tests > TAU() 1`] = `
"
┌───┬───────────────────┐
│   │ A                 │
├───┼───────────────────┤
│ 1 │ 6.283185307179586 │
└───┴───────────────────┘
"
`;

exports[`Example tests > TRUNC(1) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 1 │
└───┴───┘
"
`;
